Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+struct HTTPRequestStateMachine {
+    fileprivate enum State {
+        case initialized
+        case running(RequestState, ResponseState)
+        case finished
+
+        case failed(Error)
+    }
+
+    fileprivate enum RequestState {
+        enum ExpectedBody {
+            case length(Int)
+            case stream
+        }
+
+        enum ProducerControlState: Equatable {
```

Comment:
Why not just make this type `:String`, and you can avoid the `CustomStringConvertible` below?

Reply:
Bump